File: problems/890/solution.go
package main
func findAndReplacePattern(words []string, pattern string) []string {
    ans, patSignature := []string{}, getStringSignature(pattern)
    
    for _, word := range words {
        if patSignature == getStringSignature(word) {
            ans = append(ans, word)
        }
    }
    
    return ans
}

func getStringSignature(str string) string {
    ans, freq := "", map[rune]int{}
    
    for _, ch := range str {
        if _, ok := freq[ch]; !ok {
            freq[ch] = len(freq) + 1
        }
        
        ans += string('a' + freq[ch])
    }
    
    return ans
}

File: problems/890/solution.cpp
class Solution {
public:
    vector<string> findAndReplacePattern(vector<string>& words, string pattern) {
        vector<string> ans;
        vector<char> patFreqSorted = getCharsFreqMapSorted(pattern);
        
        for (const string& word: words) {
            if (patFreqSorted == getCharsFreqMapSorted(word)) {
                ans.push_back(word);
            }
        }
        
        return ans;
    }
    
    vector<char> getCharsFreqMapSorted(const string& str) {
        vector<char> ans;
        unordered_map<char, unsigned short int> um;
        
        for (const char& ch: str) {
            if (not um.count(ch)) {
                um[ch] = um.size() + 1;
            }
            
            ans.push_back('a' + um[ch]);
        }
        
        return ans;
    }
};